web/internal/metadata.bzl

```diff
@@ -0,0 +1,153 @@
+# Copyright 2016 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A library of functions for working with web_test metadata files."""
+
+load('//web/internal:collections.bzl', 'lists')
+load('//web/internal:files.bzl', 'files')
+
+
+def _merge_files(ctx, merger, output, inputs):
+  """Takes a list of input metadata files, and produces a merged output file.
+
+  Args:
+    ctx: a skylark rule context
+    merger: the WTL metadata merger executable
+    output: a File object for the output file
+    inputs: a list of File objects for the input files
+  """
+  paths = [i.path for i in inputs]
+  short_paths = [i.short_path for i in inputs]
+  args = ['--output', output.path] + paths
+
+  ctx.action(
+      outputs=[output],
+      inputs=inputs,
+      executable=merger,
+      arguments=args,
+      mnemonic='METADATAMERGER',
+      progress_message='merging %s' % (', '.join(short_paths)))
+
+
+def _create_file(ctx,
+                 output,
+                 capabilities=None,
+                 environment=None,
+                 browser_label=None,
+                 test_label=None,
+                 web_test_files=None):
+  """Generates a web_test metadata file with specified contents."""
+  content = '{\n  "_comment": "generated file for %s"' % ctx.label
+  if capabilities:
+    content += ',\n  "capabilities": ' + capabilities
+  if environment:
+    content += ',\n  "environment": "' + environment + '"'
+  if browser_label:
+    content += ',\n  "browserLabel": "%s"' % browser_label
+  if test_label:
+    content += ',\n  "testLabel": "%s"' % test_label
+  if web_test_files:
+    content += ',\n  "webTestFiles": ['
+    first = True
+    for f in web_test_files:
+      if first:
+        first = False
+      else:
+        content += ','
+      content += '\n' + _web_test_files_to_string(ctx, f, '    ')
+    content += '\n  ]'
+  content += '\n}\n'
+
+  ctx.file_action(output=output, content=content, executable=False)
+
+
+def _get_files(ctx, attr_names):
+  """Finds all of the web_test_metadata files in given attributes.
+
+  It ensures that all web_test_metadata files position in the returned list is
+  based on the last attribute in attr_names that they appear in.
+
+  Args:
+    ctx: the context object for this rule.
+    attr_names: the names of attributes to get web_test_metadata files from.
+  Returns:
+    a list of all of the web_test_metadata files in the specificied attributes.
+  """
+  metadata_files = []
```

Comment:
Out of curiosity, why are you using list as your go-to data structure? You almost certainly want [set](https://www.bazel.io/versions/master/docs/skylark/lib/set.html). It does everything you want and more. It's much more advanced than `set()` in Python. Rather than being a hash table, it's a tree of arrays. This data structure is built efficiently and iteratively across rules. It can maintain different types of ordering. It's awesome and should make it so you no longer need the list library.